modules/javafx.graphics/src/main/java/com/sun/javafx/css/ImmutablePseudoClassSetsCache.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.sun.javafx.css;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import javafx.css.PseudoClass;
+
+/**
+ * A cache for immutable sets of {@link PseudoClass}es.
+ */
+public class ImmutablePseudoClassSetsCache {
+    private static final Map<Set<PseudoClass>, Set<PseudoClass>> CACHE = new HashMap<>();
+
+    /**
+     * Returns an immutable set of {@link PseudoClass}es.
+     * <p>
+     * Note: this method may or may not return the same instance for the same set of
+     * {@link PseudoClass}es.
+     *
+     * @param pseudoClasses a set of {@link PseudoClass} to make immutable, cannot be {@code null}
+     * @return an immutable set of {@link PseudoClass}es, never {@code null}
+     * @throws NullPointerException when {@code pseudoClasses} is {@code null} or contains {@code null}s
+     */
+    public static Set<PseudoClass> of(Set<PseudoClass> pseudoClasses) {
+        Set<PseudoClass> cachedSet = CACHE.get(Objects.requireNonNull(pseudoClasses, "pseudoClasses cannot be null"));
+
+        if (cachedSet != null) {
+            return cachedSet;
+        }
+
+        Set<PseudoClass> copy = Set.copyOf(pseudoClasses);
```

Comment:
The set that is returned from this method will probably be passed into the method over and over again. Every time, `CACHE.get(...)` will call `pseudoClasses.hashCode()`, which in turn iterates over all elements to compute the hash code. Have you considered precomputing the hash code of the set, so that this repeated recalculation is not required? While this might seem like a micro-optimization, we are potentially dealing with a method that is called very often.

Here is an alternative implementation for an immutable set that precomputes and stores its hash code:
```java
final class ImmutableSet<T> extends AbstractSet<T> {
    private final T[] elements;
    private final int hashCode;

    @SuppressWarnings("unchecked")
    ImmutableSet(Set<T> elements) {
        this.elements = (T[])elements.toArray();
        this.hashCode = elements.hashCode();
    }

    @Override
    public Object[] toArray() {
        return elements.clone();
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<>() {
            int index;

            @Override
            public boolean hasNext() {
                return index < elements.length;
            }

            @Override
            public T next() {
                try {
                    return elements[index++];
                } catch (ArrayIndexOutOfBoundsException ignored) {
                    throw new NoSuchElementException();
                }
            }
        };
    }

    @Override
    public int size() {
        return elements.length;
    }

    @Override
    public int hashCode() {
        return hashCode;
    }
}
```

Reply:
we can even consider to calculate it lazily, even we want to take this approach

Reply:
These sets are still created on demand, just as often as before, the main change is that there will no longer be thousands of exact duplicate sets in memory that can't be shared because they are modifiable.  The passed in sets will in 99% of the cases be newly allocated sets outside of our control, allocated just nanoseconds before.  The hash code therefore will still need to be computed in almost all cases.  These are then compared with the keys in the map, which already have a pre-computed hash code (courtesy of `HashMap`).

It would be exceedingly rare (and probably a mistake) to pass in a set that you got from the cache.  If you use the cache, you know you have an immutable version, and so can share it freely.  Methods that expose these again should be documented that the set is already immutable (see `Match#getPseudoClasses` for example).  Aside from loading new style sheets, the code that will be doing the most calls to `ImmutablePseudoClassSetsCache#of` is in the `CssStyleHelper`:

                PseudoClassState pseudoClassState = new PseudoClassState();

                pseudoClassState.addAll(parent.pseudoClassStates);
                pseudoClassState.retainAll(parent.styleHelper.triggerStates);

                retainedStates[count++] = ImmutablePseudoClassSetsCache.of(pseudoClassState);

As you can see, it constructs a new mutable set, and converts it to an immutable one.  We can't avoid its hash code calculation. 

Also, the returned immutable set is currently a highly optimized implementation from `ImmutableCollections` (although they don't cache the hash code). They for example have memory storage requirements similar to an array, but still offer `O(1)` contains checks thanks to an open addressing hashing scheme that uses probing.  Even that barely matters though as most of these sets will contain only a couple of elements (most of them 1, some of them 2, and extremely rarely 3 or more).  Still, we'd be trading one optimization for another and a new maintenance burden.

Reply:
I see. If the cached sets are not passed into the cache again, it doesn't seem to be all that useful to precompute the hash.